dbms/src/Interpreters/SortSpillContext.cpp

```diff
@@ -31,11 +31,33 @@ bool SortSpillContext::updateRevocableMemory(Int64 new_value)
     if (!in_spillable_stage)
         return false;
     revocable_memory = new_value;
-    if (enable_spill && operator_spill_threshold > 0 && revocable_memory > static_cast<Int64>(operator_spill_threshold))
+    if (auto_spill_status == AutoSpillStatus::NEED_AUTO_SPILL || (enable_spill && operator_spill_threshold > 0 && revocable_memory > static_cast<Int64>(operator_spill_threshold)))
     {
         revocable_memory = 0;
         return true;
     }
     return false;
 }
+
+Int64 SortSpillContext::triggerSpill(Int64 expected_released_memories)
+{
+    if (!in_spillable_stage || !isSpillEnabled())
+        return expected_released_memories;
+    auto total_revocable_memory = getTotalRevocableMemory();
+    if (total_revocable_memory >= MIN_SPILL_THRESHOLD)
+    {
+        AutoSpillStatus old_value = AutoSpillStatus::NO_NEED_AUTO_SPILL;
+        if (auto_spill_status.compare_exchange_strong(old_value, AutoSpillStatus::NEED_AUTO_SPILL))
+        {
+            expected_released_memories = std::max(expected_released_memories - total_revocable_memory, 0);
```

Comment:
What if updateRevocableMemory is called concurrently at this time?

Reply:
If `auto_spill_status.compare_exchange_strong` is done, it will trigger spill, otherwise, it will be triggered to spill in the next iterator.

Reply:
Ok, maybe `revocable_memory` needs to be set to 0 in `finishOneSpill`?
Consider a case where `updateRevocableMemory` sets `revocable_memory` to a large value after `triggerSpill` sets `revocable_memory` to 0.
Then after `finishOneSpill`, `revocable_memory` is still a large value, but there are no blocks to spill.
```
1. triggerSpill  set revocable_memory to 0
2. updateRevocableMemory set revocable_memory  to big_value
3. finishOneSpill revocable_memory  is big_value but blocks is empty.
4. AutoTrigger will call again because of the wrong big_value of revocable_memory  
```

Reply:
ok